### ISSUE TITLE
cmd/compile: make optimized functions' parameter stack layout available to debuggers

### ISSUE THREAD

**Issue by heschik** [golang/go#27039]
To call a function that takes arguments (and to read the result) debuggers need to be able to form a stack frame for the function. With knowledge of the argument types and their order, this is pretty straightforward. In an unoptimized program, our DWARF exposes that information in two ways: first, the formal_parameter DIEs are in stack order, so you can infer the layout, and second, they have their home stack slot offsets as their locations.

In an optimized function, neither of these is usable; `putPrunedScopes` reorders the parameter DIEs, and the location lists aren't reliable enough to use to form the stack frame.

I'm not sure what the right way forward is. We could fix one or both of the above, or introduce a new custom attribute that defines the "canonical" stack slot for a parameter.

Retain DIE order: I don't have a clear sense of why the vars need to be sorted; @thanm will probably know more here. It's a little bit unsatisfying to rely on DIE ordering though.

Improve location lists: This will be very hard to get right without simply giving arguments special treatment. For example, the current location list generation logic will not notice unused arguments at all. It should be straightforward to generate location entries for each parameter at function entry, though, and we could just do that.

New custom attribute: easy, but really feels like giving up. I'd personally rather leave this as a last resort.

cc @dr2chase, @aarzilli, @derekparker

**Comment by heschik** [comment]
I agree, but for reasons like the unused argument issue I mentioned above, I expect it will be difficult to get the location lists completely reliable. (Practically, there are also quite a few accuracy issues with them too, so it would be a long time coming no matter what.) But special handling for parameters shouldn't be too hard. I'm leaning slightly in that direction.

**Comment by cherrymui** [comment]
Couldn't the parameter stack layout be derived from the type of the function?

**Comment by cherrymui** [comment]
I was just thinking what gdb's `ptype` uses. It seems it is the formal parameter DIEs, and as you mentioned, the order is sometimes incorrect. Maybe it makes sense to make the the formal parameter DIEs to have the correct order, so `ptype` would also work. (I don't know how difficult it is, though.)

**Comment by heschik** [comment]
@cherrymui's point led me to do what I really should've done before I filed this bug -- read the spec.

> 3.3.4 Declarations Owned by Subroutines and Entry Points
>
> The declarations enclosed by a subroutine or entry point are represented by debugging information entries that are owned by the subroutine or entry point entry. **Entries representing the formal parameters of the subroutine or entry point appear in the same order as the corresponding declarations in the source program.**

So we are actually in violation of the spec.

**Comment by derekparker** [comment]
@heschik can we get input from @thanm on why the parameters are being sorted in an optimized function.

**Comment by heschik** [comment]
He's OOO for a couple days, I'm sure we'll get it sorted out (rimshot) soon though.

**Comment by thanm** [comment]
[Back from vacation, now catching up on email]

Sorting: I added this to make things easier for generation of DWARF inline info DIEs.  At the point where DWARF is generated we're seeing a post-optimization version of the function (or to be more accurate, several combined functions due to inlining), each inlined body has to be matched up to pre-optimization decl list (not as easy as you might thing, since vars can be deleted, split into pieces, renamed, etc). The original version of the code just looked at the source position info for the variables to determine declaration order, but that proved to be problematic (front end was sometimes assigning the same location to two different vars). 

I could certain make an attempt at refactoring things to get rid of the sorting if that would be helpful.

**Comment by derekparker** [comment]
Thanks @thanm.

**Comment by heschik** [comment]
I may have forgotten some context, but the change above is sufficient, right?
